### ISSUE TITLE
[BC break v2.6] SingleSignOnLoginRequestSubscriber breaks Password reset

### ISSUE THREAD

**Issue by spackmat** [sulu/sulu#7447]
| Q | A
| --- | ---
| Sulu Version | 2.6.1 (since 2.6.0)
| PHP Version | all
| DB Version | all
| Browser Version | all

# Actual Behavior

After upgrading to Sulu 2.6 the password reset stopped working. I investigated and found out that the new `SingleSignOnLoginRequestSubscriber` from the `SecurityBundle` prevents the `sendEmailAction` of the `ResettingController` to be called, when no `singleSignOnAdapter` is found. When I comment out that two lines setting another Response and call `stopPropagation()`, the reset controller is called again.

```` php
        $adapter = $this->singleSignOnAdapterProvider->getAdapterByDomain($domain);
        if (!$adapter) {
            if ($password) {
                return;
            }

            $event->setResponse(new JsonResponse(['method' => 'json_login'], 200));
            $event->stopPropagation();

            return;
        }
````

# Expected Behavior

The password reset should work when the SSO is not enabled. Edit: And also for a local account with an e-mail address for which no SSO adapter is configured in a mixed SSO/local-user scenario.

# Steps to Reproduce

Have Sulu 2.6 installed and try to send a password reset email and observe that its controller code is not executed at all and instead of an empty response, get `{method: "json_login"}` that comes from the ``SingleSignOnLoginRequestSubscriber` aborting the process.

# Possible Solutions

~~Only register that `SingleSignOnLoginRequestSubscriber` if the SSO is used or~~ fix the `SingleSignOnLoginRequestSubscriber` so it does allow the `sendEmailAction` to be executed, when the SSO is not enabled or a  reset for a local account is requested.

**Comment by spackmat** [comment]
And BTW I came across that problem, because the password reset ignores the `framework.mailer.headers.From` setting and only uses the `sulu_security.reset_password.mail.sender` setting or falls back to `'no-reply@' . $request->getHost()`. That should also not be the case.

**Comment by alexander-schranz** [comment]
> The password reset should work independently of the new SSO.

No, that would open a Security whole as that would workaround the single sign on provider. If Single sign on is enabled also password reset need to be done on your SSO page as they need validate if the user still valid and so Sulu Admin does not allow to reset the passwords (as it also don't require a passsword) when SSO is enabled and instead the `SingleSignOnLoginRequestSubscriber` should redirect you to the SingleSignOnProvider depending on your SSO providers page you can or cannot reset there the password, but that part is out of hand for Sulu. You can still change the password inside Sulu admin if a password passed authentication is used on the website or somewhere else where you use the Sulu User instance.

**Comment by spackmat** [comment]
But I don't have SSO enabled and then the password reset should work or removed at all. For now, it does not work since the `SingleSignOnLoginRequestSubscriber` is always loaded and so the reset is never callable.

So sorry for my misleading sentence "The password reset should work independently of the new SSO.", I changed that to "The password reset should work when SSO is not enabled."

Edit: So I think the correct snippet for what should be achieved would be:

```` php
        $adapter = $this->singleSignOnAdapterProvider->getAdapterByDomain($domain);
        if (!$adapter) {
            // without that  `|| $isResetPassword` the password reset feature was always disabled
            // in this case explicitly when we were in a non SSO context
            if ($password || $isResetPassword) {
                return;
            }

            $event->setResponse(new JsonResponse(['method' => 'json_login'], 200));
            $event->stopPropagation();

            return;
        }
````

Since when there is no `$adapter` found for the user's `$domain` we are not in an SSO context and so the reset should be allowed.

**Comment by alexander-schranz** [comment]
That is strange if there is no SSO enabled it should not even load that service and skip the registration of `single_sign_on.xml` file: https://github.com/sulu/sulu/blob/a57671127bde0227699c4e609052a514ecba9c98/src/Sulu/Bundle/SecurityBundle/DependencyInjection/SuluSecurityExtension.php#L116

**Comment by spackmat** [comment]
@alexander-schranz maybe that is because also in the `services.xml` of the SecurityBundle, the `SingleSignOnLoginRequestSubscriber` is loaded as `sulu_security.open_id_login_subscriber`?

Anyhow: In my case, it is loaded and executed and I don't have any `single_sign_on` config set. And my proposed fix seems to be helpful as I consider the current code a bug. Inside the no-adapter-branch we are not in n SSO context for the user in question, so the request should not be cancelled.

**Comment by alexander-schranz** [comment]
I see I think there was something wrong in a rebase and accidently they were kept in both files. Does all work like expected if you remove the single sign on services from the services.xml and don't load no single_sign_on services if no providers are configured. https://github.com/sulu/sulu/blob/a57671127bde0227699c4e609052a514ecba9c98/src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml#L392-L426 ?

**Comment by spackmat** [comment]
I'm back in the office next week, but when the `SingleSignOnLoginRequestSubscriber` is not loaded, it should work.

But that does not fix the problem, as the `SingleSignOnLoginRequestSubscriber` is implemented in a way that it can handle a mixed scenario of SSO accounts and standard accounts, but that implementation has a bug. When the `SingleSignOnLoginRequestSubscriber` is removed completely and with SSO enabled the normal accounts should be disabled, then the code paths for handling the mixed scenario is obsolete and should be removed at all. Or just fix it with my provided fix to repair the whole thing, so that a accitentally enabled `SingleSignOnLoginRequestSubscriber` does not break the password reset.

**Comment by spackmat** [comment]
> Does all work like expected if you remove the single sign on services from the services.xml and don't load no single_sign_on services if no providers are configured.

I now tested that (remove the `sulu_security.open_id_login_subscriber` definition from the `sulu/src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml`) and that also works. Not quite surprising, but that does not fix the problem with the bug in the `SingleSignOnLoginRequestSubscriber`: In a mixed scenario the password reset then still never works and always fails silently (i.e. without an error message to the users nor a log entry). And when a mixed scenario with SSO and normal login is not intended, then the whole code paths within `SingleSignOnLoginRequestSubscriber` supporting that do not make sense anyhow.

**Comment by alexander-schranz** [comment]
Nice do you want for the first part a pull request. The other part we may need have a deeper look to check about the mixed mode, but atleast the existing installation should be fixed then.

**Comment by spackmat** [comment]
Hey, I didn't want to contribute to Sulu anymore and besides that, I don't have any clue what side effects I would trigger by removing the service definition of `sulu_security.open_id_login_subscriber` and dont't want to take the responsibility for such a PR. 

As I said, from my point of view that is nonetheless more of a symptom of the actual underlying bug inside the `SingleSignOnLoginRequestSubscriber` that should be fixed. But I did not author that code and I'm not interested in any more investigation work on the problem, so please take my report and my already invested investigation and my proposed solution and do whatever you internally do with such bug reports. I don't want to discuss that any further as I already wasted too many hours on that thing.

**Comment by rogamoore** [comment]
I do not understand the whole topic because I'm not using SSO at the moment, but do I understand correctly, that a quick fix would be
> remove the sulu_security.open_id_login_subscriber definition from the sulu/src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml

?

In that case I'm happy to do a quick PR

**Comment by alexander-schranz** [comment]
@rogamoore Yes that should atleast fix the first part.

**Comment by rogamoore** [comment]
Done, let's see if the tests will be green.

**Comment by alexander-schranz** [comment]
Reopened as we need to have a look at the other part when SSO is enabled and used in mixed mode.
